### ISSUE TITLE
Cross Origin Server Push

### ISSUE THREAD

**Issue by worenga** [h2o/h2o#1284]
Given a certificate that contains Subject Alternative Names for two domains A and B,
Is it possible to push resources for domain B for a request from domain A with h2o?

**Comment by deweerdt** [comment]
AFAIK, H2O doesn't currently push requests for domains that don't match the authority of the request. This code: https://github.com/h2o/h2o/blob/master/lib/core/util.c#L607 effectively checks that authorities match before deciding whether to add the request to the pushed list or not.

**Comment by worenga** [comment]
@deweerdt you are correct. I removed the check and daisy-chained the authority together with the path to push_path and it seems to work (at least for my fastcgi handler).

Looking at the code I am wondering whether the decision to block cross-origin pushes was done intentionally, e.g., due to security concerns or is this something which could land in future versions?

**Comment by deweerdt** [comment]
> Looking at the code I am wondering whether the decision to block cross-origin pushes was done intentionally, e.g., due to security concerns or is this something which could land in future versions?

I think the ideal case would be to have a way in H2O to validate that we are authoritative for the content we push. That would mean having access to the SAN and be able to perform the same validation that a browser would do. I think this is necessary because RFC 7540 "10.4.  Cacheability of Pushed Responses" says that "Pushed responses for which an origin server is not authoritative (see Section 10.1) MUST NOT be used or cached."

The next best thing, might be a configuration knob to turn strict checks on or off, because only the administrator of H2O knows whether we can trust the backend servers or not.

**Comment by kazuho** [comment]
It is correct that the restriction is intentional.

> I think the ideal case would be to have a way in H2O to validate that we are authoritative for the content we push. That would mean having access to the SAN and be able to perform the same validation that a browser would do. I think this is necessary because RFC 7540 "10.4. Cacheability of Pushed Responses" says that "Pushed responses for which an origin server is not authoritative (see Section 10.1) MUST NOT be used or cached."
> 
> The next best thing, might be a configuration knob to turn strict checks on or off, because only the administrator of H2O knows whether we can trust the backend servers or not.

If somebody is interested in doing work in this area, I would argue that what the person should do first is to look into if browsers respect cross-origin server pushes. Unless they do, doing cross-origin pushes will be just waste of bandwidth.

**Comment by worenga** [comment]
In my experiments using chromium 59 did respect cross-origin push. I'll have to check other browsers.

I think whitelisting authorities in the hostconf would be a good solution.

**Comment by ConnorKrammer** [comment]
Chrome allows cross-origin push and so does Safari (though Safari had a few other bugs, last I checked). I haven't tested Firefox, but it supports the ORIGIN frame as per [RFC 8336](https://tools.ietf.org/html/rfc8336), and I see that there's an [open pull request](https://github.com/h2o/h2o/pull/1199) for that feature included here as a v2.3 milestone.

Jake Archibald has an [article](https://jakearchibald.com/2017/h2-push-tougher-than-i-thought/) on the details of HTTP/2 push—it's almost a year old now, but it covers the cross-domain points and includes some compatibility notes.

One really good use-case for cross-origin push is pushing content along with a 3xx redirect, which saves a round trip and basically eliminates the redirect latency. ([Ctrl blog](https://www.ctrl.blog/entry/http2-push-redirects) describes this well.) Unfortunately, I recently ran into a roadblock while attempting to implement this for no-www to www redirects in Nginx: because you can't push www assets from a no-www domain even on the same connection and from the same IP address, the browser will never claim pushed resources from the cache during the redirect. For the same reason, you can't push static assets hosted on a subdomain, which is a desirable use-case.

**Comment by worenga** [comment]
I think you need an certificate, which includes the certificates you want to push stuff for as a SAN. Moreover, at least for chrome i remember that the browser might still perform DNS lookups for both origins, so they should match.

**Comment by tdtgit** [comment]
Any update on this when #1199 is merged?
